app/Util/Builders.hs

```diff
@@ -32,16 +34,25 @@ buildWidgetString =
             (summary n)
             (body n)
             (hintString n)
+            (actionString n)
     )
     ""
 
 buildHintString :: Map Text Variant -> String
-buildHintString = Data.Map.foldrWithKey (\k v s -> s ++ showEntry k v) ""
+buildHintString = buildJsonString
+
+buildJsonString :: Map Text Variant -> String
+buildJsonString pairs = "[" ++ (Data.List.intercalate ", " (Data.Map.foldrWithKey (\k v s -> s ++ showEntry k v) [] pairs)) ++ "]"
```

Comment:
I think for hints it would be better to expose them in the form 

```
{
   "key1": "value1",
   "key2": "value2",
   ...
}
```
I think that you usually know which hints to expect, so accessing them as `end-hints["key"]` would be the most convenient. What do you think?

Reply:
Oh you're probably right about that... but at that point, we might conclude to actually get rid of the whole widget by hint configuration stuff all together, and just expose a json array of notifications instead of yuck code. Conditionals in eww expression are sad though, because it's ternaries only, and having to nest ternaries is really sad.
```
(for notification in end-notifications
   (literal :content { notification.hints?.battery != null ? "(battery-notification :notification ${notification})" : "(notification :notification ${notification})" }))
```
that being said, this is pretty nice. Doesn't lend it's self well to configuring timeouts based on hints though.

Reply:
That is a really good point. However, I would prefer to minimize the amount of boilerplate yuck you have to write, and it seems that such a for loop would always be needed. But exposing the data of a notification as one big JSON object seems much nicer than having 7 different parameters, most of which you probably wont use. So an end notification widget should look like this
```
(defwidget end-notification 
 [notification]
 ... your content ...)
```
instead of
```
(defwidget end-notification 
 [end-id end-appname end-appicon end-summary end-body end-hints]
 ... your content ...)
```

Simplifying the custom notification logic is tempting, and in fact the current solution is neither here nor there. As mentioned in issue #6, you often don't want to distinguish notifications based on the `end-type` hint, but on things like the application name. The way to do this right now is to use ternary expressions and literals, so why should the `end-type` hint be any special?

The question is *how* to unify this. In #6, I suggested to introduce small conditional expressions in the config, because manually working with literal widgets isn't great (due to missing syntax checking & highlighting). However, this has the possibility to blow up the scope and introduce a lot of complexity. Also, you can always extract most of your literal yuck code into a new widget and then simply call that from the literal block (like you did in your example). Of course the ugly nested ternaries stay and there is no way to define custom timeouts. 

I'm leaning towards your approach of getting rid of custom notifications altogether. The custom timeouts I only added because I could, not because I actually needed them.  Do you have a use-case for them that cannot be handled by the urgency-based timeouts?

Reply:
> I'm leaning towards your approach of getting rid of custom notifications altogether. The custom timeouts I only added because I could, not because I actually needed them. Do you have a use-case for them that cannot be handled by the urgency-based timeouts?

Well I do like battery notifications having infinite timeout, but they can just be critical urgency anyway. So yeah, honestly I don't.